nixos/modules/services/x11/redshift.nix

```diff
@@ -88,51 +71,101 @@ in {
       example = "/bin/redshift-gtk";
       description = ''
         Redshift executable to use within the package.
+
+        To use Gammastep, pass either <literal>/bin/gammastep</literal>
+        or <literal>/bin/gammastep-indicator</literal>.
       '';
     };
 
-    extraOptions = mkOption {
-      type = types.listOf types.str;
-      default = [];
-      example = [ "-v" "-m randr" ];
+    settings = mkOption {
+      type = types.submodule {
+        freeformType = settingsFormat.type;
+      };
+
+      default = {};
+
+      example = {
+        redshift = {
+          temp-day = 5500;
+          temp-night = 3700;
+          brightness-day = 1.0;
+          brightness-night = 0.8;
+          fade = 1;
+        };
+        randr = {
+          screen = 0;
+        };
+      };
+
       description = ''
-        Additional command-line arguments to pass to
-        <command>redshift</command>.
+        The configuration to pass to Redshift/Gammastep.
+
+        Available options for Redshift described in
+        <citerefentry>
+          <refentrytitle>redshift</refentrytitle>
+          <manvolnum>1</manvolnum>
+        </citerefentry> and for Gammastep in
+        <citerefentry>
+          <refentrytitle>gammastep</refentrytitle>
+          <manvolnum>1</manvolnum>
+        </citerefentry>.
       '';
     };
   };
 
-  config = mkIf cfg.enable {
+  config = let
+    mainSection = if strings.hasInfix "gammastep" cfg.executable
+      # https://gitlab.com/chinstrap/gammastep/-/commit/1608ed61154cc652b087e85c4ce6125643e76e2f
+      then "general"
+      else "redshift";
+    target = if (cfg.settings.${mainSection}.adjustment-method or null) == "drm"
+      then "basic.target"
+      else "graphical-session.target";
+    configFile = settingsFormat.generate "redshift.conf" cfg.settings;
+    fullExecutable = "${cfg.package}${cfg.executable}";
+  in mkIf cfg.enable {
     # needed so that .desktop files are installed, which geoclue cares about
     environment.systemPackages = [ cfg.package ];
 
-    services.geoclue2.appConfig.redshift = {
+    services.geoclue2.appConfig.redshift = mkIf isGeoclue2 {
       isAllowed = true;
       isSystem = true;
     };
 
-    systemd.user.services.redshift =
-    let
-      providerString = if lcfg.provider == "manual"
-        then "${toString lcfg.latitude}:${toString lcfg.longitude}"
-        else lcfg.provider;
-    in
-    {
+    assertions = [
+      {
+        assertion = (cfg.settings ? "redshift") != (cfg.settings ? "general");
+        message = ''Both [redshift] and [general] sections found in settings. For gammastep use "general = {}" instead of "redshift = {}" in "server.redshift.settings".'';
+      }
+      {
+        assertion = (pathExists fullExecutable);
+        message = ''Executable not found. Are you using "/bin/redshift" for "pkgs.redshift" or "/bin/gammastep" for "pkgs.gammastep"?'';
+      }
```

Comment:
This should be removed, because checking whether a derivation path exists can only be done after the derivation is built, meaning in order to evaluate such a NixOS configuration, Nix needs to first evaluate to this point, then build the derivation, then evaluate again, which should be avoided because it's slow.